### ISSUE TITLE
Access the `.fetch` response directly to check for `response.ok` ?

### ISSUE THREAD

**Issue by oheydrew** [alex-cory/use-http#112]
Hi there! 

Great work on the library- It looks to have almost all of what we need, but I've been playing with it and I've found an edge-case that I'd love some help with.

In our current implementations of our own hand-rolled `useFetch`, we have something that checks for any `response.status` that is not in the `200-299` range, ie: `!response.ok`:

```
const isOK = response => {
  if (!response.ok) {
    throw response
  }

  return response
}

const toJSON = response => response.json()

return fetch(`${config.BASE_API}/${endpoint}`, finalHeaders)
         .then(isOK)
         .then(toJSON)
```

This is so that we can basically call anything that's not a "200" an "error", as far as our front-end is concerned. We've got cases in our backend where we'd be receiving 422's, or 401's as responses, and since `fetch` doesn't error on those by default, we check for `response.ok`

Unfortunately due to the fact that `use-http` does all the response parsing and checking internally, by the time I get to the response data in the interceptors, I've already lost the `response.ok` & `response.status` objects etc.

Is there a way around this with the current implementation? It'd be nice if there was a way to specify interceptions even earlier, for this kind of situation, so that our `422`'s return to the `error` object rather than the `data` one.

Alternatively, happy to discuss other ways to solve this issue, even if I'm going about this kind of wrong. 

Thanks!

**Comment by alex-cory** [comment]
Btw, I'm in Western Australia at the moment. Noticed you're in Melbourne haha.

Regarding this issue, hmmm... I will take a look into it right now

**Comment by alex-cory** [comment]
This is a bug! Thank you for spotting this. I will fix it right now

**Comment by alex-cory** [comment]
Are you looking for a way to `setError` in the `interceptors.response`? Maybe I change the syntax to look something like
```jsx
const App = () => {
  const options = {
    interceptors: {
      response: ({ setError, response }) => {
        if (!response.ok) setError('some error you want')
      }
    }
  }
  return <Provider url='some url' options={options}><RestOfApp /></Provider>
}
```

**Comment by alex-cory** [comment]
Found the bug, should have a fix out in the next hour or so

**Comment by oheydrew** [comment]
Hey Alex,

Haha, yes, Melbourne indeed! I didn't realise this was a home-grown library :) thanks for your hard work. 

It's an interesting problem to solve. I wouldn't want to change the API so much that the returned payload changes, or that the interceptors people currently have are affected, so it's a tricky thing to figure through, in regards to whether the response is decoded or not when the interceptor starts, and indeed what is returned. 

I tend to think the first solution might be best, but you'd need to make sure it still parses the response, otherwise the API may change...?

**Comment by oheydrew** [comment]
Alternatively, an option to implement your own "res.ok?" checks internally would be cool. Either default behaviour or as a flag that could be set... Not sure which way you'd want to go. Part of me thinks that 422's and 401's shouldn't be coming back as "data", but as "error", but I know that differs from fetch's default spec.

**Comment by alex-cory** [comment]
So in this update, the 422's and 401's you can check to see in the `interceptors.response` and `throw` if that happens. i.e.
```jsx
const App = () => {
  const options = {
    interceptors: {
      response: ({ setError, response }) => {
        if ([422, 402].includes(response.status)) throw Error('some error you want')
        if (!response.ok) throw Error('some error you want')
      }
    }
  }
  return <Provider url='some url' options={options}><RestOfApp /></Provider>
}

const RestOfApp = () => {
  const { loading, error } = useFetch({ onMount: true })
  return (
    <>
      {error && error.message} {/* <-- this error will be set if a 422 or 401 happens */}
    </>
  )
}
```

**Comment by alex-cory** [comment]
The only problem here is that `Response` can't be spread or anything. So I can't do
```js
const res = new Response()
const customRes = { data: 'cool', ...res }
// customRes === { data: 'cool' }
```
but, you can modify the `Response` directly
```js
```js
const res = new Response()
res.data = 'cool'
// res == Response({ ok: true, data: 'cool', ...etc })
```
```

**Comment by alex-cory** [comment]
If you're curious, about what I did to fix this, [check out this PR](https://github.com/alex-cory/use-http/pull/113/files).

**Comment by alex-cory** [comment]
This is fixed, tested, and published in `v0.1.91`.

**Comment by oheydrew** [comment]
Thanks @alex-cory - Superb turnaround! I went home for the day and I can try it out this morning 😂 Nice one. I'll let you know how it goes :)
